coreax/data.py

```diff
@@ -25,42 +25,83 @@
 
 
 @overload
-def _atleast_2d_consistent(array: Shaped[Array, " n d"]) -> Shaped[Array, " n d"]: ...
+def _atleast_2d_consistent(
+    arrays: Shaped[Array, " n d *p"],
+) -> Shaped[Array, " n d *p"]: ...
+
+
+@overload
+def _atleast_2d_consistent(  # pyright:ignore[reportOverlappingOverload]
+    arrays: Sequence[Shaped[Array, " n d *p"]],
+) -> List[Shaped[Array, " n d *p"]]: ...
 
 
 @overload
 def _atleast_2d_consistent(  # pyright:ignore[reportOverlappingOverload]
-    array: Shaped[Array, " d"],
+    arrays: Shaped[Array, " d"],
 ) -> Shaped[Array, " d 1"]: ...
 
 
 @overload
-def _atleast_2d_consistent(  # pyright:ignore[reportOverlappingOverload],
-    array: Union[Shaped[Array, ""], float, int],
-) -> Shaped[Array, " 1 1"]: ...
+def _atleast_2d_consistent(  # pyright:ignore[reportOverlappingOverload]
+    arrays: Sequence[Shaped[Array, " d"]],
+) -> List[Shaped[Array, " d 1"]]: ...
+
+
+@overload
+def _atleast_2d_consistent(  # pyright:ignore[reportOverlappingOverload]
+    arrays: Shaped[Array, " d"],
+) -> Shaped[Array, " d 1"]: ...
+
+
+@overload
+def _atleast_2d_consistent(  # pyright:ignore[reportOverlappingOverload]
+    arrays: Sequence[Shaped[Array, ""]],
+) -> List[Shaped[Array, " 1 1"]]: ...
```

Comment:
_Also_ also - I think that at present, `[torch.zeros(2,3), torch.zeros(3,4)]` might not pass the type hint, since the _number_ of dimensions is the same for each array in the list, but the _size_ of each dimension is different. You'll need to annotate the array dimensions with `_` to fix this - I think we unavoidably lose some specificity in our type hints here.

Alternatively, we could just require that only one array is passed in, and leave it to client code to handle conversion of multiple arrays at once. This has the benefit of allowing our type hints to be more accurate, but does have the obvious disadvantage of requiring more verbose client code when converting multiple arrays. However, I don't think we're actually using `_atleast_2d_consistent` to convert multiple arrays at once anywhere, so I don't think this would be an issue.

Reply:
> Also, this new set of overloads has removed the ability to pass in a `float` or `int` - was this intentional?

It was, this matches the upcoming implementation in JaxNumpy of `atleast_2d` (https://github.com/google/jax/blob/main/jax/_src/numpy/lax_numpy.py#L3626-L3635) that states they will no longer accept non-array inputs. The code still works but I think we should avoid encouraging its use in this way,

Reply:
> _Also_ also - I think that at present, `[torch.zeros(2,3), torch.zeros(3,4)]` might not pass the type hint, since the _number_ of dimensions is the same for each array in the list, but the _size_ of each dimension is different. You'll need to annotate the array dimensions with `_` to fix this - I think we unavoidably lose some specificity in our type hints here.
> 
> Alternatively, we could just require that only one array is passed in, and leave it to client code to handle conversion of multiple arrays at once. This has the benefit of allowing our type hints to be more accurate, but does have the obvious disadvantage of requiring more verbose client code when converting multiple arrays. However, I don't think we're actually using `_atleast_2d_consistent` to convert multiple arrays at once anywhere, so I don't think this would be an issue.

We do use it to convert multiple arrays in the algorithms implemented in the other repo so I would like to keep this implementation if possible. 

I like the use of `_` more, will add unless you have other thoughts?

Reply:
> To be super pedantic, I think we also need other overloads here for the case where we pass in a list of arrays of differing dimensions - for example, the input `[torch.zeros(2), torch.zeros(2,2)]` doesn't match any of the given overloads.

Yes I agree, I could add one `Sequence[Union[...]]` overload to replace all the `Sequence` ones that includes all the possible dimensions?

Reply:
Or to pass a typechecker we could add a more general `Array` -> `Array` overload, and comment it to cover the cases you describe?

Reply:
I am conscious if we try add an overload for every possible combination of array shapes then we will be approaching 10+ overloads

Reply:
I think just adding one `Sequence[Union[Shaped[...], Shaped[...], ...]]` overload is good enough, using `_` for the dimension annotations. Then we only have to deal with 4 overoads (3 array -> array, 1 sequence -> sequence).

Reply:
@rg936672 done this now

Reply:
Need to fix the tests for this function ! One second